train.py

```diff
@@ -261,148 +264,299 @@ def build_graph(reader,
+      self.last_log = 0
+
+      try:
+        logging.info("%s: Entering training loop.", task_as_string(self.task))
+        while not sv.should_stop():
+
+          batch_start_time = time.time()
+          _, global_step_val, loss_val, predictions_val, labels_val = sess.run(
+              [train_op, global_step, loss, predictions, labels])
+          seconds_per_batch = time.time() - batch_start_time
+
+          self.now = time.time()
+          is_time_to_save = (self.now - self.last_save) > _SAVE_INTERVAL_SECONDS
+          is_time_to_log = (self.now - self.last_log) > _LOG_INTERVAL_SECONDS
+          should_save = self.is_master and is_time_to_save and self.train_dir
+          should_log = is_time_to_log or should_save
+
+          if should_log or should_save:
+            examples_per_second = labels_val.shape[0] / seconds_per_batch
+            hit_at_one = eval_util.calculate_hit_at_one(predictions_val,
+                                                        labels_val)
+            perr = eval_util.calculate_precision_at_equal_recall_rate(
+                predictions_val, labels_val)
+            gap = eval_util.calculate_gap(predictions_val, labels_val)
+
+          if should_log:
+            logging.info(
+                "%s: training step " + str(global_step_val) + "| Hit@1: " +
+                ("%.2f" % hit_at_one) + " PERR: " + ("%.2f" % perr) + " GAP: " +
+                ("%.2f" % gap) + " Loss: " + str(loss_val),
+                task_as_string(self.task))
+            self.last_log = self.now
+
+          if should_save:
+            logging.info("%s: Writing summary.", task_as_string(self.task))
+            sv.summary_writer.add_summary(
+                utils.MakeSummary("model/Training_Hit@1", hit_at_one),
+                global_step_val)
+            sv.summary_writer.add_summary(
+                utils.MakeSummary("model/Training_Perr", perr), global_step_val)
+            sv.summary_writer.add_summary(
+                utils.MakeSummary("model/Training_GAP", gap), global_step_val)
+            sv.summary_writer.add_summary(
+                utils.MakeSummary("global_step/Examples/Second",
+                                  examples_per_second), global_step_val)
+            sv.summary_writer.flush()
+            self.last_save = self.now
+
+      except tf.errors.OutOfRangeError:
+        logging.info("%s: Done training -- epoch limit reached.",
+                     task_as_string(self.task))
+
+    logging.info("%s: Exited training loop.", task_as_string(self.task))
+    sv.Stop()
+    return hit_at_one, perr
+
+  def start_server_if_distributed(self):
+    """Starts a server if the execution is distributed."""
+
+    if self.cluster:
+      logging.info("%s: Starting trainer within cluster %s.",
+                   task_as_string(self.task), self.cluster.as_dict())
+      server = start_server(self.cluster, self.task)
+      target = server.target
+      device_fn = tf.train.replica_device_setter(
```

Comment:
Added the parameter '"ps_device="/job:ps"'. It is the default value but it seems that being explicit can improve readability. 

Removing the parameter "merge_devices=True". It is also the default but this parameter is on the path to deprecation and specifying "merge_devices=False" triggers a warning.